src/compiler/core/src/commonMain/kotlin/community/flock/wirespec/compiler/core/parse/TypeParser.kt

```diff
@@ -91,54 +114,84 @@ object TypeParser {
+        when (token.type) {
+            is LeftParentheses -> {
+                val regex = eatToken<RegExp>().bind()
+                Reference.Primitive.Type.Pattern(regex.value).also {
+                    eatToken<RightParentheses>().bind()
+                    eatToken().bind()
+                }
+            }
+            else -> null
+        }
+    }
+
+    fun TokenProvider.parseWirespecType() = either {
+        val wirespecType = when (val type = token.type) {
+            is WsString -> {
+                eatToken().bind()
                 Reference.Primitive(
-                    type = Reference.Primitive.Type.String,
-                    isNullable = it,
+                    isNullable = isNullable().bind(),
+                    type = Reference.Primitive.Type.String(
+                        pattern = parseWirespecTypePattern().bind(),
+                    ),
                 )
             }
 
-            is WsBytes -> { it ->
+            is WsBytes -> {
+                eatToken().bind()
                 Reference.Primitive(
                     type = Reference.Primitive.Type.Bytes,
-                    isNullable = it,
+                    isNullable = isNullable().bind(),
                 )
             }
 
-            is WsInteger -> { it ->
+            is WsInteger -> {
+                eatToken().bind()
                 Reference.Primitive(
-                    type = Reference.Primitive.Type.Integer(type.precision.toPrimitivePrecision()),
-                    isNullable = it,
+                    type = Reference.Primitive.Type.Integer(
+                        precision = type.precision.toPrimitivePrecision(),
+                    ),
+                    isNullable = isNullable().bind(),
                 )
             }
 
-            is WsNumber -> { it ->
+            is WsNumber -> {
+                eatToken().bind()
                 Reference.Primitive(
                     type = Reference.Primitive.Type.Number(type.precision.toPrimitivePrecision()),
-                    isNullable = it,
+                    isNullable = isNullable().bind(),
                 )
             }
 
-            is WsBoolean -> { it ->
+            is WsBoolean -> {
+                eatToken().bind()
                 Reference.Primitive(
                     type = Reference.Primitive.Type.Boolean,
-                    isNullable = it,
+                    isNullable = isNullable().bind(),
                 )
             }
 
-            is WsUnit -> Reference::Unit
+            is WsUnit -> {
+                eatToken().bind()
+                Reference.Unit(
+                    isNullable = isNullable().bind(),
+                )
+            }
 
-            is TypeIdentifier -> { it ->
-                current.shouldBeDefined().bind()
+            is TypeIdentifier -> {
+                token.shouldBeDefined().bind()
+                val value = token.value
+                eatToken().bind()
```

Comment:
all these eatToken().bind() statements can be removed if `val previous` suggestion is accepted